### PR TITLE
fix: can't compile in MSYS2/Mingw64 on windows 10 64bits

### LINKED ISSUE #837: can't compile in MSYS2/Mingw64 on windows 10 64bits
**Describe the bug**
can't compile in MSYS2/Mingw64 on windows 10 64bits

**To Reproduce**
```bash
git clone --depth 1 https://github.com/arturo-lang/arturo
cd arturo
nim ./build.nims install mini
```
then I read
```
=====================================================================
                               _
                              | |
                     __ _ _ __| |_ _   _ _ __ ___
                    / _` | '__| __| | | | '__/ _ \
                   | (_| | |  | |_| |_| | | | (_) |
                    \__,_|_|   \__|\__,_|_|  \___/

                     Arturo Programming Language
                      (c)2022 Yanis Zafirópulos

=====================================================================
 ► INSTALL
=====================================================================

--------------------------------------------
 ● Checking environment...
--------------------------------------------
   os: windows
   compiler: Nim v1.7.3

--------------------------------------------
 ● Building...
--------------------------------------------
   version: 0.9.80 b/3195
   config: @mini

```
however there is no arturo.exe on my harddisk

**Expected behavior**
arturo.exe is built and placed in ~/.arturo

### PR DESCRIPTION
**Pull request by RickBarretto**
# Description

I read [this issue](https://github.com/arturo-lang/arturo/issues/837) and tried to build Arturo in my machine, but I had an error, as I said [here](https://github.com/arturo-lang/arturo/issues/837#issuecomment-1339761080).
I tried a lot of ways to fix it, and I finally found one.

This PR changes just one line inside of `if hostOS=="windows"`, so It must not change anything for the others Operational Systems.

This was the way I found to build locally on my machine.

> Fixes #837

As I said, it's working on my local machine. So I'll wait for CI.
Thanks @retsyo for report this issue!

### Build Info

```
--------------------------------------------
 ��� Checking environment...
--------------------------------------------
   os: windows
   compiler: Nim v1.6.6

--------------------------------------------
 ��� Building...
--------------------------------------------
   version: 0.9.80 b/3298
   config: @mini
   flags: --skipUserCfg:on --colors:off -d:danger --panics:off --mm:orc -d:useMalloc --checks:off -d:ssl --cincludes:extras --opt:speed --nimcache:.cache --path:src   -d:NOASCIIDECODE -d:NOCLIPBOARD -d:NODIALOGS -d:NOGMP -d:NOPARSERS -d:NOSQLITE
-d:NOWEBVIEW -d:MINI
```
And
```
--------------------------------------------
 ��� Checking environment...
--------------------------------------------
   os: windows
   compiler: Nim v1.6.6

--------------------------------------------
 ��� Building...
--------------------------------------------
   version: 0.9.80 b/3298
   config: @full
   flags: --skipUserCfg:on --colors:off -d:danger --panics:off --mm:orc -d:useMalloc --checks:off -d:ssl --cincludes:extras --opt:speed --nimcache:.cache --path:src
```

As you can see, I'm building on 0.9.80 b/3298 version, using nim 1.6.6

### Additional Notes

I built it on VsCodium that was not reading `PKG_CONFIG_PATH`, so none needs to be changed with this variable.
Other times ago on Msys2, I tried to change the `PKG_CONFIG_PATH` to right path and then build. `pkg-config` command was working, but the build was not, giving me [this same error](https://github.com/arturo-lang/arturo/issues/837#issuecomment-1339761080).

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Comment by retsyo** [comment]
yes, `arturo.exe` can be created now

one more thing, if I use
```
/e/prg/nim/nim-1.6.10/bin/nim ./build.nims install mini
```
then `libwinpthread-1.dll` is need to run `arturo.exe`. However if I use
```
/e/prg/nim/nim-1.6.10/bin/nim ./build.nims install
```
then `libwinpthread-1.dll` is not need

I know that the dependency `libwinpthread-1.dll` is caused by GCC. but should it be 
consistent to use `libwinpthread-1.dll` or not no matter `/e/prg/nim/nim-1.6.10/bin/nim ./build.nims install` or `/e/prg/nim/nim-1.6.10/bin/nim ./build.nims install mini`

**Comment by retsyo** [comment]
p.s.

on windows, some dll files are need, can we compile all them into ONE arturo.exe?
```
2022/12/09  13:20         4,250,016 arturo.exe
2022/12/08  14:37         3,450,752 libcrypto-1_1-x64.dll
2022/11/26  10:01           955,034 libgmp-10.dll
2022/12/08  14:37           697,728 libssl-1_1-x64.dll
2022/11/21  09:15         1,918,464 sqlite3_64.dll
2022/12/08  09:08           857,600 webview.dll
2022/12/08  14:37           137,648 WebView2Loader.dll
```

**Review by drkameleon — APPROVED** [review]
Awesome job! I wouldn't have spotted this one... no matter how hard I tried! 😉

**Comment by drkameleon** [comment]
> yes, `arturo.exe` can be created now
> 
> one more thing, if I use
> 
> ```
> /e/prg/nim/nim-1.6.10/bin/nim ./build.nims install mini
> ```
> 
> then `libwinpthread-1.dll` is need to run `arturo.exe`. However if I use
> 
> ```
> /e/prg/nim/nim-1.6.10/bin/nim ./build.nims install
> ```
> 
> then `libwinpthread-1.dll` is not need
> 
> I know that the dependency `libwinpthread-1.dll` is caused by GCC. but should it be consistent to use `libwinpthread-1.dll` or not no matter `/e/prg/nim/nim-1.6.10/bin/nim ./build.nims install` or `/e/prg/nim/nim-1.6.10/bin/nim ./build.nims install mini`

Very interesting that you mention it.

To be perfectly honest, I thought none of the 2 versions needed *libthread* the way things are right now. But I guess I was wrong... I'll look into it.

And yes, it doesn't make any sense that the MINI builds needs it, while the other one doesn't.

**Comment by drkameleon** [comment]
> p.s.
> 
> on windows, some dll files are need, can we compile all them into ONE arturo.exe?
> 
> ```
> 2022/12/09  13:20         4,250,016 arturo.exe
> 2022/12/08  14:37         3,450,752 libcrypto-1_1-x64.dll
> 2022/11/26  10:01           955,034 libgmp-10.dll
> 2022/12/08  14:37           697,728 libssl-1_1-x64.dll
> 2022/11/21  09:15         1,918,464 sqlite3_64.dll
> 2022/12/08  09:08           857,600 webview.dll
> 2022/12/08  14:37           137,648 WebView2Loader.dll
> ```

This is a tricky one, although what you say *does* make sense.

What I can do (and I'll open an issue for that as a matter of fact, so that we can keep track of it) is try to pack *some* of the DLL dependencies in the same binary for Windows. So, that - at least - we *reduce* the number of DLL dependencies.

However, I'm not really sure it will be possible in all cases:

- The Webview-related dependencies (for the FULL builds) are complicated in themselves. I may be able to make it happen but I'm not sure how...
- The GMP dependency is a totally tough one: basically, the GMP is GPL-licensed and - correct me if I'm wrong - but I don't think a non-GPL'ed project can statically link it. Which is obviously a deal breaker. The problem here is... if we do need BigNum support out of the box, this one (GMP) is by far the most performant library available. What could be a solution to this - which is something that I have been brainstorming tbh - is to offer alternative versions of the binaries (pretty much like with MINI vs FULL), but also for different types of features (e.g. GMP-powered BigNum support vs non-GMP BigNum's - which is not gonna be thaaaat easy, but I guess it's feasible). Of course, I would really won't to offer TOO many binaries, because, in the end, people will end up being more confused than happy... I think.

I don't know. I'm all ears to hear what you think! 😉

**Comment by drkameleon** [comment]
I have opened 2 new issues for the topics mentioned by @retsyo, so that we can keep track of them in a more efficient way and we can obviously continue any further related discussion there.

That being said, given that @RickBarretto's fix works like a charm (again: great stuff! keep it coming!! 😉 ), I'm closing this one.

-----

So... ready to merge! 🚀

**Comment by RickBarretto** [comment]
So, there is many problems with statically link Arturo with these DLLs. These're problems with Licenses. 

> TL;DR: Arturo is a MIT licensed software, and these DLLs have different licenses, what sometimes can give troubles to the project. The main example are GPL and LGPL that are considered viral licenses. **I'm not a lawyer, so I can't give any advice here, but I collected some information.**

Firstly, Arturo is under [MIT License](https://choosealicense.com/licenses/mit/) [ [source](https://github.com/arturo-lang/arturo/blob/master/LICENSE) ]. Save this information.

Secondly, let's see what are the Libraries licenses

## Libraries' Licenses
- [gmp 10](https://gmplib.org/) is under [LGPL v3](https://www.gnu.org/licenses/lgpl-3.0.html) and [GPL v2](https://www.gnu.org/licenses/old-licenses/gpl-2.0.html) since version 6. [ [source](https://gmplib.org/manual/Copying) ]
  - read the Licenses to know the problems with
- [crypto 1.1](https://www.openssl.org/docs/man3.0/man7/crypto.html) and [openssl 1.1](https://www.openssl.org/docs/man1.1.1/man7/ssl.html) are under [dual Open SSL and SSLeay Licenses](https://www.openssl.org/source/license-openssl-ssleay.txt). [ [source](https://www.openssl.org/docs/man3.0/man7/crypto.html#COPYRIGHT) ]
  - read [this from Legal Questions](https://www.openssl.org/docs/faq.html#LEGAL1), it may be important
  - If I'm not wrong Arturo is current using 1.1 version of both. [For 3.0 version, OpenSSL is migrating for Apache license](https://www.openssl.org/docs/man3.0/man7/migration_guide.html#License-Change).
- [sqlite 3](https://www.sqlite.org/index.html) is under [Public Domain](https://wikiless.org/wiki/Public_Domain?lang=en). [ [source](https://www.sqlite.org/copyright.html) ]
  - read the section *"Warranty of Title"* to know the problems with. Sometimes you'll need to [buy a sqlite license](https://www.sqlite.org/purchase/license?).
  - Additional readings about troubles using this license: [Public Domain Trouble Spots | Stanford](https://fairuse.stanford.edu/overview/public-domain/trouble-spots/), [Validity of the Creative Commons Zero 1.0 Universal Public Domain Dedication and its usability for bibliographic metadata from the perspective of German Copyright Law ](https://rd-alliance.org/sites/default/files/cc0-analysis-kreuzer.pdf)
- And, if `WebView2Loader.dll` is what I'm think, it is the [WebView2 from Microsoft](https://learn.microsoft.com/en-us/microsoft-edge/webview2/). As you can see, under [Files to ship with the app Topic](https://learn.microsoft.com/en-us/microsoft-edge/webview2/concepts/distribution#files-to-ship-with-the-app), you'll find this same dll. It says that you can use both, statically and dynamically. It's under ["liberal BSD-style license"](https://github.com/MicrosoftEdge/WebView2Feedback/issues/240#issuecomment-639134153), I think that [this is the License the author of the last commentary's link was referring to | Nuget](https://www.nuget.org/packages/Microsoft.Web.WebView2/0.9.430/License). 

That said, let's see what this Licenses implies:

## Troubles

LGPL

> "Generally, dynamic linking of LGPL code is considered best practice, as static linking makes meeting the license requirements more complicated. While it is possible to comply with LGPL code that is integrated into proprietary code as a statically linked library, it requires more effort. There is a kind of safe harbor for using LGPL code as a dynamically linked library; [...]"
> [Open Source Software Licenses 101: The LGPL License](https://fossa.com/blog/open-source-software-licenses-101-lgpl-license/)

> "[...] using a dynamically loaded LGPL module is fine even in proprietary software."
> [@ncw](https://github.com/ncw/gmp/issues/9#issuecomment-399385300)

You can use but dynamically linked if I'm not wrong. Never try to statically link it, because it's a viral license. So, be careful.

Open SSL and SSLeay

> "For information on intellectual property rights, please consult a lawyer. The OpenSSL team does not offer legal advice. [...]"
> [Do I need patent licenses to use OpenSSL?](https://www.openssl.org/docs/faq.html#LEGAL1)

I don't know what are the implications of statically link this software. Anyway, [here is the license](https://www.openssl.org/source/license-openssl-ssleay.txt).

SQLite's Public Domain

> "The core SQLite database software is in the [public domain](https://sqlite.org/copyright.html). No license is necessary to use SQLite or to build SQLite into your products. However, licenses are available to satisfy the following needs:
> - You want indemnity against claims of copyright infringement.
> - You are using SQLite in a jurisdiction that does not recognize the public domain.
> - You are using SQLite in a jurisdiction that does not recognize the right of authors to dedicate their work to the public domain.
> - You want to hold a tangible legal document as evidence that you have the legal right to use and distribute SQLite.
> - Your legal department tells you that you have to purchase a license."
> [License And Warranty Of Title](https://www.sqlite.org/purchase/license?)

So, there is some problems with Public Domain. Please, read the two references: [Public Domain Trouble Spots | Stanford](https://fairuse.stanford.edu/overview/public-domain/trouble-spots/), [Validity of the Creative Commons Zero 1.0 Universal Public Domain Dedication and its usability for bibliographic metadata from the perspective of German Copyright Law ](https://rd-alliance.org/sites/default/files/cc0-analysis-kreuzer.pdf). It'll explain more.

---
Note, I know that issues were opened about this problem. But I prefer to post it here, I'll write a comment referring to this commentary. And also, I'm not a lawyer, so I'm not on position to give legal advices.

**Comment by celtic-coder** [comment]
An excellent write-up, @RickBarretto! 👍

**Comment by RickBarretto** [comment]
Thank you!!

**Comment by RickBarretto** [comment]
But, @retsyo it would be a awesome idea if it were not for the licenses.

**Comment by drkameleon** [comment]
@RickBarretto First of all, great job there! It's very interesting and practical to have all the information in one place!! 😉 

**Regarding the licenses:**

- **GPL/LGPL**: As much as I respect RMS's work and contributions, this is probably the number 1 license I disagree with, even ideologically lol. I mean it's a sure-fire way to lock you in. So, yes, I generally avoid them as much as I can. So, why pick a GPL/LGPL-licensed project, like GMP, for Arturo? 'Coz it's simply the best one out there, if we want the fastest BigNum processing available. It's a very complicated project, with TONS of work-hours invested in it throughout the years and every single project that depends on efficient BigNum calculations and handling *does* depends on it. Could we create an alternative implementation? (which is something that I have thought of, as an alternative - or as an alternative, e.g. for MINI builds) Sure, we could. As a matter of fact, there is already quite some work done in this aspect in Nim as well! The problem is... this is never gonna be as fast as GMP. But I would be open to the idea of having our own, GPL/GMP-independent BigNum support - at least, as an 
- alternative option.

- **OpenSSL**: Regarding these two libraries (namely `libcrypto` & `libssl`), I think we'd have no issue static-linking them whatsoever. At least, from a licensing perspective. I think my main issue was technical, hence I decided to put it off. I may be wrong but I think Fabio (@h3rald) already static-links to them in [Min](https://github.com/h3rald/min), so I guess we could... copy a few ideas regarding this part 😄 

- **Public domain**: Honestly, it had never occured to me that using (or static-linking, for this matter) a public domain project could be source of troubles. I'll have a look into it. Sure thing is, with more DB types being supported, at some point we will have to depend on other libraries as well... So, perhaps that's why I haven't already even attempted to static-link SQLite, but we'll see about that I guess...

Again, as @celtic-coder already said, excellent write-up and a great list to have around for reference! 🚀

**Comment by retsyo** [comment]
wait, I download the zipped src, now both `/e/prg/nim/nim-1.6.10/bin/nim ./build.nims install` and `/e/prg/nim/nim-1.6.10/bin/nim ./build.nims install mini` need `libwinpthread-1.dll`

**Comment by drkameleon** [comment]
> wait, I download the zipped src, now both `/e/prg/nim/nim-1.6.10/bin/nim ./build.nims install` and `/e/prg/nim/nim-1.6.10/bin/nim ./build.nims install mini` need `libwinpthread-1.dll`

That's super weird!

First of all, you mean the *zip* from the CI builds, right?

To make sure we're on the same page and since I'm not very familiar with windows, is there any - easy - way to tell which are a binary's DLLs dependencies on Windows? (perhaps, I could include something relevant in the CI builds and see what is going on...)

**Comment by dumblob** [comment]
> * **GPL/LGPL**: As much as I respect RMS's work and contributions, this is probably the number 1 license I disagree with, even ideologically lol. I mean it's a sure-fire way to lock you in. So, yes, I generally avoid them as much as I can. So, why pick a GPL/LGPL-licensed project, like GMP, for Arturo? 'Coz it's simply the best one out there, if we want the fastest BigNum processing available. It's a very complicated project, with TONS of work-hours invested in it throughout the years and every single project that depends on efficient BigNum calculations and handling _does_ depends on it. Could we create an alternative implementation? (which is something that I have thought of, as an alternative - or as an alternative, e.g. for MINI builds) Sure, we could. As a matter of fact, there is already quite some work done in this aspect in Nim as well! The problem is... this is never gonna be as fast as GMP. But I would be open to the idea of having our own, GPL/GMP-independent BigNum support - at least, as an alternative option.

Statically linking (L)GPL libraries to MIT software is not an issue iff both MIT and (L)GPL are being complied to (independent from each other).

> * **OpenSSL**: Regarding these two libraries (namely `libcrypto` & `libssl`), I think we'd have no issue static-linking them whatsoever. At least, from a licensing perspective. I think my main issue was technical, hence I decided to put it off. I may be wrong but I think Fabio (@h3rald) already static-links to them in his [**Min programming language interpreter**](https://github.com/h3rald/min) (also written in Nim _and_ MIT-licensed) - so, I guess we could... copy a few ideas regarding this part smile

Exactly - there should not be any issues with statically linking. Again, OpenSSL license(s) have to be complied to independently from other licenses.

> * **Public domain**: Honestly, it had never occured to me that using (or static-linking, for this matter) a public domain project could be source of troubles. I'll have a look into it. Sure thing is, with more DB types being supported, at some point we will have to depend on other libraries as well... So, perhaps that's why I haven't already even attempted to static-link SQLite, but we'll see about that I guess...

Fedora has had this very same issue with public domain many years ago. They solved it simply be relicensing to MIT (because, you know, public domain allows that :wink:).
